agent/consul/rpc.go

```diff
@@ -521,6 +522,23 @@ func (s *Server) blockingQuery(queryOpts structs.QueryOptionsCompat, queryMeta s
 	var timeout *time.Timer
 	var queryTimeout time.Duration
 
+	// REVIEW: The spec for this placed the atomic inc for queries_blocking inside the RUN_QUERY statement
+	// however, there are a combination of factors that I believe should mean we should place it before.
+	// - `goto RUN_QUERY` on L545 above means that we'd instrument a non-blocking query as blocking for its duration.
+	// - using defer dec on L537 would mean that we inc the counter multiple times as the blocking query waits, while
+	//   we only dec on the return of blockingQuery(). We'd need swap out defer for a dec on retry on L621 as well as
+	// 	 on the return at L625
+	// -> There I think it's simpler if we don't instrument the RUN_QUERY statement and use the
+	// 	  start and end of s.blockingQuery() as the points of measurement.
+
+	// atomic inc our gauge of blockingQueries
+	atomic.AddUint64(&s.queriesBlocking, 1)
+	// atomic dec when we return from blockingQuery()
+	defer atomic.AddUint64(&s.queriesBlocking, ^uint64(0))
+	// set gauge directly to a float32 of our Server's queriesBlocking
+	queriesBlocking := float32(atomic.LoadUint64(&s.queriesBlocking))
```

Comment:
`atomic.AddUint64` returns the new value after the atomic op so we can save making a separate atomic update here.

So you could save the extra atomic operation with:

```go
queriesBlocking := atomic.AddUint64(&s.queriesBlocking, 1)
defer atomic.AddUint64(&s.queriesBlocking, ^uint64(0))
metrics.SetGauge([]string{"rpc", "queries_blocking"}, float32(queriesBlocking))
```

Reply:
Ah, missed that in the api docs -- fixed in latest revision